crates/aptos-crypto/src/bls12381.rs

```diff
@@ -0,0 +1,263 @@
+//! // example, we assume the aggregator receives a signature share from every third signer (for simplicity).
+//!
+//! // Here, we simulate the aggregator receiving some signature shares.
+//! let mut sigshares_received = vec![];
+//! for sigshare in sigshares.into_iter().step_by(3) {
+//!     sigshares_received.push(sigshare);
+//! }
+//!
+//! // Here, the aggregator aggregates the received signature shares into a multisignature.
+//! let multisig = BLS12381Signature::aggregate(sigshares_received.clone()).unwrap();
+//!
+//! // Third, the aggregator checks that the _optimistic_ aggregation from above succeeded by
+//! // verifying the multisig. For this, the aggregator will need to know the public keys of the
+//! // signers whose signature shares were aggregated, so that it can aggregate them.
+//! let mut pubkeys_to_agg = vec![];
+//! for kp in key_pairs.iter().step_by(3) {
+//!     pubkeys_to_agg.push(&kp.public_key);
+//! }
+//!
+//! let aggpk = BLS12381PublicKey::aggregate(pubkeys_to_agg.clone()).unwrap();
+//!
+//! // Lastly, the aggregator checks the aggregated multisig verifies successfully.
+//! assert!(multisig.verify(&message, &aggpk).is_ok());
+//!
+//! // If the multisig failed verification, the aggregator can individually verify each of the
+//! // signature shares to identify which ones are invalid and exclude them. There are also optimized
+//! // methods for identifying bad signature shares faster when their relative frequency is low [^LM07].
+//! // However, we will not implement these yet.
+//! for (sigshare, pk) in zip(sigshares_received, pubkeys_to_agg) {
+//!     assert!(sigshare.verify(&message, &pk).is_ok());
+//! }
+//! ```
+//!
+//! References:
+//! [^Bold03]: Threshold Signatures, Multisignatures and Blind Signatures Based on the Gap-Diffie-Hellman-Group Signature Scheme; by Boldyreva, Alexandra; in PKC 2003; 2002
+//! [^BLS04]: Short Signatures from the Weil Pairing; by Boneh, Dan and Lynn, Ben and Shacham, Hovav; in Journal of Cryptology; 2004; https://doi.org/10.1007/s00145-004-0314-9
+//! [^LM07]: Finding Invalid Signatures in Pairing-Based Batches; by Law, Laurie and Matt, Brian J.; in Cryptography and Coding; 2007
+//! [^MOR01]: Accountable-Subgroup Multisignatures: Extended Abstract; by Micali, Silvio and Ohta, Kazuo and Reyzin, Leonid; in Proceedings of the 8th ACM Conference on Computer and Communications Security; 2001; https://doi-org.libproxy.mit.edu/10.1145/501983.502017
+//! [^RY07]: The Power of Proofs-of-Possession: Securing Multiparty Signatures against Rogue-Key Attacks; by Ristenpart, Thomas and Yilek, Scott; in Advances in Cryptology - EUROCRYPT 2007; 2007
+
+/// Domain separation tag (DST) for hashing a message before signing it.
+const DST_BLS_MULTISIG_IN_G2_WITH_POP: &[u8] = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_";
+
+pub mod bls12381_pop;
+pub mod bls12381_keys;
+pub mod bls12381_sigs;
+
+#[cfg(test)]
+mod tests {
```

Comment:
it's weird to put tests in this file, you probably want to put them into a separate unit test under unit_tests dir